### PR TITLE
Functional Tests - Add tests for categories

### PR DESCRIPTION
**Pull request by nesrineabdmouleh**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Adding Functional tests for categories
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | TEST_PATH="functional/BO/catalog/categories/*" URL_FO=SHOP_URL npm run specific-test

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/15988)

### CODE REVIEW

File: tests/puppeteer/campaigns/data/faker/category.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+const faker = require('faker');
+
+const groupAccess = ['Visitor', 'Guest', 'Customer'];
+
+module.exports = class Category {
+  constructor(categoryToCreate = {}) {
+    this.name = categoryToCreate.name || faker.commerce.department();
+    this.displayed = categoryToCreate.displayed || 'Yes';
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/data/faker/category.js]: Please replace this with Boolean, and how to check that in functions too

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+  });
+  after(async () => {
+    await helper.closeBrowser(browser);
+  });
+  // Login into BO and go to categories page
+  loginCommon.loginBO();
+  it('should go to "Catalog>Categories" page', async function () {
+    await this.pageObjects.boBasePage.goToSubMenu(
+      this.pageObjects.boBasePage.productsParentLink,
+      this.pageObjects.boBasePage.categoriesLink);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: Make sure that we know where the line end (like that)
```suggestion
      this.pageObjects.boBasePage.categoriesLink
   );
```

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    await expect(pageTitle).to.contains(this.pageObjects.categoriesPage.pageTitle);
+  });
+  it('should reset all filters and get Number of Categories in BO', async function () {
+    if (await this.pageObjects.categoriesPage.elementVisible(
+      this.pageObjects.categoriesPage.filterResetButton,
+      2000)) {
+      await this.pageObjects.categoriesPage.resetFilter();
+    }
+    numberOfCategories = await this.pageObjects.categoriesPage.getNumberFromText(
+      this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    await expect(numberOfCategories).to.be.above(0);
+  });
+  // 1 : Filter Categories with all inputs and selects in grid table
+  describe('Filter Categories', async () => {
+    it('should filter by Id \'9\'', async function () {
+      await this.pageObjects.categoriesPage.filterCategories(
+        'input',
+        'id_category',
+        Categories.art.id);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+      const numberOfCategoriesAfterFilter = await this.pageObjects.categoriesPage.getNumberFromText(
+        this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+      await expect(numberOfCategoriesAfterFilter).to.be.at.most(numberOfCategories);
+      /* eslint-disable no-await-in-loop */
+      for (let i = 1; i <= numberOfCategoriesAfterFilter; i++) {
+        const textColumn = await this.pageObjects.categoriesPage.getTextContent(
+          this.pageObjects.categoriesPage.categoriesListTableColumn.replace('%ROW', i).replace(
+            '%COLUMN',
+            'id_category'),
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: Maybe all replaces in one line

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        );
+        await expect(textColumn).to.contains(Categories.art.id);
+      }
+      /* eslint-enable no-await-in-loop */
+    });
+    it('should reset all filters', async function () {
+      await this.pageObjects.categoriesPage.resetFilter();
+      const numberOfCategoriesAfterReset = await this.pageObjects.categoriesPage.getNumberFromText(
+        this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+      await expect(numberOfCategoriesAfterReset).to.equal(numberOfCategories);
+    });
+    it('should filter by Name \'Accessories\'', async function () {
+      await this.pageObjects.categoriesPage.filterCategories(
+        'input',
+        'name',
+        Categories.accessories.name,
+      );
+      const numberOfCategoriesAfterFilter = await this.pageObjects.categoriesPage.getNumberFromText(
+        this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+          this.pageObjects.categoriesPage.categoriesListTableColumn.replace('%ROW', i).replace('%COLUMN', 'name'),
+        );
+        await expect(textColumn).to.contains(Categories.accessories.name);
+      }
+      /* eslint-enable no-await-in-loop */
+    });
+    it('should reset all filters', async function () {
+      await this.pageObjects.categoriesPage.resetFilter();
+      const numberOfCategoriesAfterReset = await this.pageObjects.categoriesPage.getNumberFromText(
+        this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+      await expect(numberOfCategoriesAfterReset).to.equal(numberOfCategories);
+    });
+    it('should filter by Description', async function () {
+      await this.pageObjects.categoriesPage.filterCategories(
+        'input',
+        'description',
+        Categories.accessories.description,
+      );
+      const numberOfCategoriesAfterFilter = await this.pageObjects.categoriesPage.getNumberFromText(
+        this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+            .replace('%ROW', i).replace('%COLUMN', 'description'),
+        );
+        await expect(textColumn).to.contains(Categories.accessories.description);
+      }
+      /* eslint-enable no-await-in-loop */
+    });
+    it('should reset all filters', async function () {
+      await this.pageObjects.categoriesPage.resetFilter();
+      const numberOfCategoriesAfterReset = await this.pageObjects.categoriesPage.getNumberFromText(
+        this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here

File: tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+      await expect(numberOfCategoriesAfterReset).to.equal(numberOfCategories);
+    });
+    it('should filter by Position \'3\'', async function () {
+      await this.pageObjects.categoriesPage.filterCategories(
+        'input',
+        'position',
+        Categories.art.position,
+      );
+      const numberOfCategoriesAfterFilter = await this.pageObjects.categoriesPage.getNumberFromText(
+        this.pageObjects.categoriesPage.categoryGridTitle);
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/catalog/categories/01_filterAndQuickEditCategory.js]: same here and in all the cases like that please

File: tests/puppeteer/campaigns/data/demo/categories.js [hunk 13]
```diff
@@ -1,7 +1,32 @@
 module.exports = {
   Categories: {
-    accessories: {id: '6', name: 'Accessories'},
-    stationery: {id: '7', name: 'Stationery'},
-    men: {id: '4', name: 'Men'},
+    art: {
+      id: '9',
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/data/demo/categories.js]:
```suggestion
      id: 9,
```
Consider id as int

File: tests/puppeteer/campaigns/data/demo/categories.js [hunk 14]
```diff
@@ -1,7 +1,32 @@
...
+      name: 'art',
+      description: 'Framed poster and vector images',
+      position: '3',
+      displayed: 'Yes',
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/data/demo/categories.js]: Consider this kind of property as boolean
```suggestion
      displayed: true,
```

File: tests/puppeteer/campaigns/data/demo/categories.js [hunk 15]
```diff
@@ -1,7 +1,32 @@
 module.exports = {
   Categories: {
-    accessories: {id: '6', name: 'Accessories'},
-    stationery: {id: '7', name: 'Stationery'},
-    men: {id: '4', name: 'Men'},
+    art: {
+      id: '9',
+      name: 'art',
+      description: 'Framed poster and vector images',
+      position: '3',
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/data/demo/categories.js]: same for position, it should be an int

File: tests/puppeteer/campaigns/data/demo/categories.js (hunk 13 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/campaigns/data/demo/categories.js]: Thank you ;) done

File: tests/puppeteer/pages/BO/BObasePage.js [hunk 16]
```diff
@@ -136,4 +141,29 @@ module.exports = class BOBasePage extends CommonPage {
...
     }
   }
+
+  /**
+   * Generate an image then upload it
+   * @param selector
+   * @param imageName
+   * @return {Promise<void>}
+   */
+  async GenerateAndUploadImage(selector, imageName) {
```
Comment (Progi1984) [inline tests/puppeteer/pages/BO/BObasePage.js]:
```suggestion
  async generateAndUploadImage(selector, imageName) {
```

File: tests/puppeteer/pages/BO/addCategory.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+   * Fill form for add/edit category
+   * @param categoryData
+   * @return {Promise<textContent>}
+   */
+  async createEditCategory(categoryData) {
+    await this.setValue(this.nameInput, categoryData.name);
+    if (categoryData.displayed) await this.page.click(this.displayed.replace('%ID', '1'));
+    else await this.page.click(this.displayed.replace('%ID', '0'));
+    await this.setValueOnTinymceInput(this.descriptionIframe, categoryData.description);
+    await this.GenerateAndUploadImage(this.categoryCoverImage, `${categoryData.name}.jpg`);
```
Comment (Progi1984) [inline tests/puppeteer/pages/BO/addCategory.js]:
```suggestion
    await this.generateAndUploadImage(this.categoryCoverImage, `${categoryData.name}.jpg`);
```

**Review by Progi1984 — COMMENTED** [review]
@SimonGrn 2 small comments :)

File: tests/puppeteer/pages/BO/BObasePage.js (hunk 16 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/pages/BO/BObasePage.js]: Thank you @Progi1984  :)

**Review by Progi1984 — APPROVED** [review]

**Pull request merged**